core/src/cluster_slot_state_verifier.rs

```diff
@@ -255,76 +332,546 @@ pub(crate) fn check_slot_agrees_with_cluster(
+                    slot: 1,
+                    bank_frozen_hash,
+                    cluster_duplicate_confirmed_hash,
+                    is_dead,
+                    is_slot_duplicate: false,
+                },
+                SlotStateUpdate::Dead,
+                vec![
+                ResultingStateChange::MarkSlotDuplicate(bank_frozen_hash),
+                ResultingStateChange::RepairDuplicateConfirmedVersion(cluster_duplicate_confirmed_hash.unwrap())],
+            )
+        },
+        dead_state_update_3: {
+            let is_dead=true;
+            // Bank hash has to be default for dead slots
+            let bank_frozen_hash = Hash::default();
+            let cluster_duplicate_confirmed_hash = Some(Hash::new_unique());
+            (
+                SlotStateHandlerArgs {
+                    slot: 1,
+                    bank_frozen_hash,
+                    cluster_duplicate_confirmed_hash,
+                    is_dead,
+                    is_slot_duplicate: true,
+                },
+                SlotStateUpdate::Dead,
+                vec![
+                ResultingStateChange::MarkSlotDuplicate(bank_frozen_hash),
+                ResultingStateChange::RepairDuplicateConfirmedVersion(cluster_duplicate_confirmed_hash.unwrap())],
+            )
+        },
+        duplicate_state_update_0: {
+            // Slot must be marked duplicate
+            let is_slot_duplicate = true;
+            // Slot is not frozen and not dead, so no action can be taken
+            let cluster_duplicate_confirmed_hash = Some(Hash::new_unique());
+            let bank_frozen_hash = Hash::default();
+            (
+                SlotStateHandlerArgs {
+                    slot: 1,
+                    bank_frozen_hash,
+                    cluster_duplicate_confirmed_hash,
+                    is_dead: false,
+                    is_slot_duplicate,
+                },
+                SlotStateUpdate::Duplicate,
+                Vec::<ResultingStateChange>::new()
+            )
+        },
+        duplicate_state_update_1: {
+            // Slot must be marked duplicate
+            let is_slot_duplicate = true;
+            // Slot is dead, so no action can be taken
+            let cluster_duplicate_confirmed_hash = Some(Hash::new_unique());
+            let bank_frozen_hash = Hash::default();
+            (
+                SlotStateHandlerArgs {
+                    slot: 1,
+                    bank_frozen_hash,
+                    cluster_duplicate_confirmed_hash,
+                    is_dead: true,
+                    is_slot_duplicate,
+                },
+                SlotStateUpdate::Duplicate,
+                vec![
+                ResultingStateChange::MarkSlotDuplicate(bank_frozen_hash)],
+            )
+        },
+        duplicate_state_update_2: {
+            // Slot must be marked duplicate
+            let is_slot_duplicate = true;
+            // Slot is frozen, so mark the duplicate
```

Comment:
nit: Slot is frozen but same hash so no action

Reply:
ah good catch, bad copy-pasta :'(